flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointSubsumeHelper.java

```diff
@@ -67,6 +75,7 @@ public static void subsume(
             }
             // Don't break out from the loop to subsume intermediate savepoints
         }
+        return lastSubsumedCheckpoint;
```

Comment:
I think it would be more clear if the latest subsumed checkpoint is recorded by the store in it's subsume action (because of the same ambiguity about what is returned from this method).

Reply:
Maybe we could record the latest subsumed checkpoint in the store, however, this might not get the core idea of current `CompletedCheckpointStore#addCheckpoint`. 
You can take a look at `DefaultCompletedCheckpointStore#addCheckpoint`, once the `checkpointStateHandleStore.addAndLock(path, checkpoint)` failed, it would trigger the subsume process anymore to ensure strict high-availability. The timing to subsume checkpoint is very important and just record a subsumed checkpoint but not known when that one is subsumed would lose information.

Reply:
WDYT about [this](https://github.com/rkhachatryan/flink/commit/10242b7aef2f4f2db94c924f940928fce4ab0fc9#diff-1bc8de30a2deacb9128e6bb8c13169b3cb692beddb08e6724f1cf2b550435effR264) approach?

Reply:
Thanks for Roman sharing the implementation of adding another method `CompletedCheckpointStore#getLatestSubsumedCheckpointID`. However, I still prefer to leverage previous `CompletedCheckpointStore#addCheckpoint` to know the last subsumed checkpoint for reasons:

1. In general, we should make the minimal changes to existing interfaces.
2. Previous method of `CompletedCheckpointStore#addCheckpoint` already act the role to subsume checkpoint/savepoint when adding a new complete checkpoint. We could achieve our goal If we just returned the subsumed checkpoint in previous method instead of return nothing.
3. Adding method of `CompletedCheckpointStore#getLatestSubsumedCheckpointID` would introduce another method and would also make us lost the timing when the checkpoint is subsumed.

After offline discussion with each other, we think maybe involving more guys could help this discussion more clear. Could you share some thoughts here? @StephanEwen @pnowojski

Reply:
edit: actually after removing field `@Nullable private CompletedCheckpoint lastSubsumedCheckpoint;` I don't see actually much difference between those two approaches. Is there a problem that I'm missing? 

Note it doesn't help that the changes are not squashed, ant that the counter proposal is also a separate commit on top of this PR. It makes reviewing and comparing both options very difficult.

Reply:
@pnowojski , thanks for your review! Sorry for late reply due to recent too busy internal work.

The reason I prefer the solution to return the checkpoint to subsume is that we might not susbume checkpoint each time especially when we adding a complete savepoint (see [FLINK-21351](https://issues.apache.org/jira/browse/FLINK-21351)). Just use a getter would lost such information.